### PR TITLE
Add out of bounds check for `offset` intrinsics

### LINKED ISSUE #1233: `offset` intrinsic doesn't check that pointers stay within bounds
The [documentation for the `offset` method](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset) mentions in the safety conditions that:

> Both the starting and resulting pointer must be either in bounds or one byte past the end of the same [allocated object](https://doc.rust-lang.org/std/ptr/index.html#allocated-object).

At present, the standard library prevents us from adding this check because otherwise we run into spurious failures (i.e., offsets that go out-of-bounds,  are frequently used).

### PR DESCRIPTION
**Pull request by celinval**
Note that the `offset` intrinsics is lowered to `Rvalue::Offset`. Thus, we have to replace the entire statement.

This PR does not address `ptr_offset_from` and `ptr_offset_from_unsigned`. I'll push those changes to a separate PR.

Resolves #1233 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by celinval** [comment]
I realized that I ended up not using the unsupported hook. I'm happy to remove it from this PR if needed.

File: library/kani_core/src/models.rs [hunk 1]
```diff
@@ -42,6 +44,71 @@ macro_rules! generate_models {
...
+            /// An offset model that checks UB.
+            #[kanitool::fn_marker = "OffsetModel"]
+            pub fn offset<T, P: Ptr<T>, O: ToISize>(ptr: P, offset: O) -> P {
+                let offset = offset.to_isize();
+                let t_size = core::mem::size_of::<T>() as isize;
+                if offset == 0 || t_size == 0 {
+                    // It's always safe to perform an offset of length 0.
+                    ptr
+                } else {
+                    let (byte_offset, overflow) = offset.overflowing_mul(t_size as isize);
```
Comment (carolynzech) [inline library/kani_core/src/models.rs:57]:
```suggestion
                    let (byte_offset, overflow) = offset.overflowing_mul(t_size);
```
You already cast to an `isize` when creating `t_size`, so I don't think this cast should be necessary.

File: library/kani_core/src/models.rs [hunk 2]
```diff
@@ -42,6 +44,71 @@ macro_rules! generate_models {
...
+                    kani::safety_check(!overflow, "Offset in bytes overflow isize");
```
Comment (carolynzech) [inline library/kani_core/src/models.rs:58]:
```suggestion
                    kani::safety_check(!overflow, "Offset in bytes overflows isize");
```

File: library/kani_core/src/mem.rs [hunk 3]
```diff
@@ -149,7 +149,14 @@ macro_rules! kani_mem {
         )]
         #[allow(clippy::not_unsafe_ptr_arg_deref)]
         pub fn same_allocation<T: ?Sized>(ptr1: *const T, ptr2: *const T) -> bool {
-            cbmc::same_allocation(ptr1, ptr2)
+            same_allocation_internal(ptr1, ptr2)
+        }
+
+        #[allow(clippy::not_unsafe_ptr_arg_deref)]
+        pub(super) fn same_allocation_internal<T: ?Sized>(ptr1: *const T, ptr2: *const T) -> bool {
```
Comment (carolynzech) [inline library/kani_core/src/mem.rs:156]: Nit: I think it would be cleaner to delete this function and just do this:

```rust
#[allow(clippy::not_unsafe_ptr_arg_deref)]
pub fn same_allocation<T: ?Sized>(ptr1: *const T, ptr2: *const T) -> bool {
  let addr1 = ptr1 as *const ();
  let addr2 = ptr2 as *const ();
  cbmc::same_allocation(addr1, addr2)
}
```
and call `same_allocation` inside of our offset model. This isn't a huge deal--it just seems simpler to keep one function, unless I'm missing something here.

File: library/kani_core/src/models.rs [hunk 4]
```diff
@@ -42,6 +44,71 @@ macro_rules! generate_models {
...
+                    let orig_ptr = ptr.to_const_ptr();
+                    let new_ptr = orig_ptr.wrapping_byte_offset(byte_offset);
+                    kani::safety_check(
+                        kani::mem::same_allocation_internal(orig_ptr, new_ptr),
+                        "Offset result and original pointer should point to the same allocation",
```
Comment (carolynzech) [inline library/kani_core/src/models.rs:63]:
```suggestion
                        "Offset result and original pointer must point to the same allocation",
```

File: tests/expected/offset-bounds-check/out_of_bounds_ub_check.expected [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Failed Checks: Offset result and original pointer should point to the same allocation
```
Comment (carolynzech) [inline tests/expected/offset-bounds-check/out_of_bounds_ub_check.expected:1]:
```suggestion
Failed Checks: Offset result and original pointer must point to the same allocation
```

File: tests/expected/offset-bounds-check/start_from_oob.expected [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Failed Checks: Offset result and original pointer should point to the same allocation
```
Comment (carolynzech) [inline tests/expected/offset-bounds-check/start_from_oob.expected:1]:
```suggestion
Failed Checks: Offset result and original pointer must point to the same allocation
```

File: tests/expected/offset-bytes-overflow/expected [hunk 7]
```diff
@@ -1,2 +1,4 @@
-FAILURE\
-offset: attempt to compute number in bytes which would overflow
\ No newline at end of file
+Failed Checks: Offset in bytes overflow isize
```
Comment (carolynzech) [inline tests/expected/offset-bytes-overflow/expected:1]:
```suggestion
Failed Checks: Offset in bytes overflows isize
```

File: tests/expected/offset-from-bytes-overflow/expected [hunk 8]
```diff
@@ -1,2 +1,4 @@
-FAILURE\
-attempt to compute number in bytes which would overflow
+Failed Checks: Offset in bytes overflow isize
```
Comment (carolynzech) [inline tests/expected/offset-from-bytes-overflow/expected:1]:
```suggestion
Failed Checks: Offset in bytes overflows isize
```

File: tests/expected/offset-invalid-args/invalid_args.rs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Check Kani output if the arguments provided to the offset intrisic are incorrect.
```
Comment (carolynzech) [inline tests/expected/offset-invalid-args/invalid_args.rs:3]:
```suggestion
//! Check Kani output if the arguments provided to the offset intrinsic are incorrect.
```

File: tests/expected/offset-overflow/expected [hunk 10]
```diff
@@ -1,2 +1,4 @@
-FAILURE\
-attempt to compute offset which would overflow
+Failed Checks: Offset result and original pointer should point to the same allocation
```
Comment (carolynzech) [inline tests/expected/offset-overflow/expected:1]:
```suggestion
Failed Checks: Offset result and original pointer must point to the same allocation
```

File: tests/expected/offset-invalid-args/invalid_args.expected [hunk 11]
```diff
@@ -0,0 +1 @@
+Cannot offset non-pointer type
```
Comment (carolynzech) [inline tests/expected/offset-invalid-args/invalid_args.expected:1]: Where does this error message come from? I would have expected to see the message from `validate_raw_ptr`.
Also, can you test that `validate_offset` catches failures too?

File: tests/expected/offset-wraps-around/main.rs [hunk 12]
```diff
@@ -1,35 +1,59 @@
 // Copyright Kani Contributors
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
-// Check that a high offset causes a "wrapping around" behavior in CBMC.
+//! This test checks that Kani correctly computes the value of a wrapping_offset in cases where
+//! the add operation wraps.
+//!
+//! Note that CBMC offset logic will wrap around the object bits, not the entire space address when
```
Comment (carolynzech) [inline tests/expected/offset-wraps-around/main.rs:7]:
```suggestion
//! Note that CBMC offset logic will wrap around the object bits, not the entire address space, when
```

File: tests/expected/offset-wraps-around/main.rs [hunk 13]
```diff
@@ -1,35 +1,59 @@
...
-        // default behavior.
-        let v_wrap: *const u128 = v_0.add(high_offset.try_into().unwrap());
-        let wrapped_offset = v_wrap.offset_from(v_0);
-        // Both offsets should be the same, but because of the "wrapping around"
-        // behavior in CBMC, `wrapped_offset` does not equal `high_offset`
+        let v_wrap: *const u128 = v_0.wrapping_add(high_offset);
+        // This should trigger UB!!
+        let wrapped_offset = unsafe { v_wrap.offset_from(v_0) };
+        // Without UB detection, the offsets are the same, but CBMC pointer arithmetic
+        // would "wrapping around" making this incorrect
```
Comment (carolynzech) [inline tests/expected/offset-wraps-around/main.rs:23]:
```suggestion
        // would "wrap around" making this incorrect
```

File: tests/expected/offset-wraps-around/main.rs [hunk 14]
```diff
@@ -1,35 +1,59 @@
...
         // https://github.com/model-checking/kani/issues/1150
         assert!(high_offset == wrapped_offset.try_into().unwrap());
     }
 }
+
+/// This harness is similar to the `original_harness`, but we replace the `offset_from` with
+/// a subtraction on the pointer addresses.
+#[kani::proof]
```
Comment (carolynzech) [inline tests/expected/offset-wraps-around/main.rs:31]: Can you update the expected file to account for these new harnesses? We should check that verification succeeds.

**Review by carolynzech — COMMENTED** [review]
Nice! I don't have as many comments as it initially appears, a good chunk of them are just propagating error message suggestions through the tests.

File: kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs [hunk 15]
```diff
@@ -205,10 +207,15 @@ impl GotocHook for Check {
 
         let (msg, reach_stmt) = gcx.codegen_reachability_check(msg, span);
 
+        // Since `cond` might have side effects, assign it to a temporary
+        // variable so that it's evaluated once, then assert and assume it
+        // TODO: I don't think `cond` can have side effects, this is MIR, it's going to be temps
+        let (tmp, decl) = gcx.decl_temp_variable(cond.typ().clone(), Some(cond), caller_loc);
```
Comment (zhassan-aws) [inline kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs:213]: Why this change?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs [hunk 16]
```diff
@@ -38,7 +38,7 @@ pub enum PropertyClass {
     /// Overflow panics that can be generated by Intrisics.
     /// NOTE: Not all uses of this are found by rust-analyzer because of the use of macros. Try grep instead.
     ///
-    /// SPECIAL BEHAVIOR: None TODO: Why should this exist?
+    /// SPECIAL BEHAVIOR: Same as SafetyCheck.
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs:41]: Yeah, I think we should get rid of them. Just not in this PR. :smile:

File: kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs [hunk 17]
```diff
@@ -747,3 +755,25 @@ impl GotocHooks {
         }
     }
 }
+
+fn check_hook(
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs:759]: yeah, I agree. I had initially moved two other hook checks to us it, but the are slightly different. I can remove it.

File: library/kani_core/src/mem.rs (hunk 3 above)
Comment (celinval) [reply library/kani_core/src/mem.rs:156]: The reason why I created a second one is because the first one is tagged with `#[unstable]` attribute. Using that one would require us to either remove the attribute or users would have to pass `-Z mem-predicates`.

File: tests/expected/offset-bounds-check/start_from_oob.expected (hunk 6 above)
Comment (celinval) [reply tests/expected/offset-bounds-check/start_from_oob.expected:1]: If the first is out of bounds, then the pointers do not point to the same allocation. Let me know if you have a different message in mind before I update all the tests.

**Review by zhassan-aws — COMMENTED** [review]
Thanks!

File: tests/expected/offset-invalid-args/invalid_args.expected (hunk 11 above)
Comment (celinval) [reply tests/expected/offset-invalid-args/invalid_args.expected:1]: So, the Rust compiler actually crashes before codegen. This is a panic message from the compiler.

I added the validation and tests just in case they eventually decide to handle this case differently, but today, the validations are redundant.

File: kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs (hunk 15 above)
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs:213]: That was a mistake. Let me revert it.

File: tests/kani/ValidValues/unaligned.rs
Comment (celinval) [inline tests/kani/ValidValues/unaligned.rs:1]: btw, I had this test in my repo for a while, and I accidentally bundled up in this PR. I hope that's OK. :smile:

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs:41]: I adjusted the message

File: library/kani_core/src/mem.rs (hunk 3 above)
Comment (carolynzech) [reply library/kani_core/src/mem.rs:156]: Ah okay, works for me then.

File: tests/expected/offset-wraps-around/main.rs (hunk 14 above)
Comment (celinval) [reply tests/expected/offset-wraps-around/main.rs:30]: Done!

**Review by carolynzech — APPROVED** [review]
